Add private key password configuration option

jatalahd opened the pull request:
- Adds an option to insert private key password in code or configuration file to avoid typing it manually every time server starts.

- With other commit in cheroot, fixes #1583

**What kind of change does this PR introduce?**
  - [ ] bug fix
  - [ X] feature
  - [ ] docs update
  - [ ] tests/coverage improvement
  - [ ] refactoring
  - [ ] other



**What is the related issue number (starting with `#`)**
#1583


**What is the current behavior?** (You can also link to an open issue here)
If ssl is used and private key has password protection,
the password needs to be entered manually every time the server starts.


**What is the new behavior (if this is a feature change)?**
With this change, the private key password can be inserted via code
or in configuration file.


**Other information**:


**Checklist**:

  - [ X] I think the code is well written
  - [ X] I wrote [good commit messages][1]
  - [ X] I have [squashed related commits together][2] after the changes have been approved
  - [ ] Unit tests for the changes exist
  - [ ] Integration tests for the changes exist (if applicable)
  - [ X] I used the same coding conventions as the rest of the project
  - [ X] The new code doesn't generate linter offenses
  - [ ] Documentation reflects the changes
  - [ X] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

jatalahd committed:
Add private key password configuration option

- Adds an option to insert private key password
  in code or configuration file to avoid typing
  it manually every time server starts.

- With other commit in cheroot, fixes #1583

webknjaz commented on cherrypy/_cpwsgi_server.py:
```diff
@@ -89,23 +89,26 @@ def __init__(self, server_adapter=cherrypy.server):
+            self.server_adapter.ssl_ciphers,
+        ]
+        ssl_kwargs = {}
+        if self.server_adapter.ssl_private_key_password is not None:
```
Why make this conditional?

webknjaz replied:
If this is meant for compatibility with older Cheroot, I think we could just bump the minimum version in the packaging metadata instead.

jatalahd replied:
Yes, it was for backwards compatibility. Now removed in the latest push.

webknjaz replied:
@jatalahd plz also bump Cheroot's minimum version to whatever introduced this interface first in `pyproject.toml`.

jatalahd replied:
Cheroot dependency version updated in latest push.

webknjaz commented on cherrypy/_cpwsgi_server.py:
```diff
@@ -89,23 +89,26 @@ def __init__(self, server_adapter=cherrypy.server):
         else:
             ssl_module = self.server_adapter.ssl_module or 'pyopenssl'
+
+        ssl_args = [
```
Can we just use kwargs for everything?

jatalahd replied:
Yes, much better that way.

webknjaz commented on cherrypy/_cpwsgi_server.py:
```diff
@@ -89,23 +89,26 @@ def __init__(self, server_adapter=cherrypy.server):
+            )
+
         if self.server_adapter.ssl_context:
             adapter_class = cheroot.server.get_ssl_adapter_class(ssl_module)
```
Perhaps, we could collapse these branches and then get rid of the additional args?
```suggestion
        set_up_tls = self.server_adapter.ssl_context or self.server_adapter.ssl_certificate
        if set_up_tls:
            adapter_class = cheroot.server.get_ssl_adapter_class(ssl_module)
            self.ssl_adapter = adapter_class(*ssl_args, **ssl_kwargs)  # FIXME: drop the indirectly unwrapped args in favor of explicit
            if self.server_adapter.ssl_context:
                self.ssl_adapter.context = self.server_adapter.ssl_context
```

jatalahd replied:
Done as suggested. There was duplicate code, so with this change it is much better.

webknjaz replied:
Marking as resolved. Will leave comments on the new diff.

webknjaz reviewed (commented):
To be mergeable, this will need a change in the core packaging metadata (min Cheroot) and some test, perhaps. Also, rebase this branch as I've posted a few fixes on `main`.

jatalahd commented:
@webknjaz; In latest push, updated the cheroot minimum version to 11.1.0. Added also one unit test for each ssl adapter, had to add pyopenssl as a test dependency. Possibly made a mistake with git when pushing, but hopefully did not mess up anything.

webknjaz commented on tox.ini:
```diff
@@ -104,6 +104,8 @@ extras =
   memcached_session
 whitelist_externals = mkdir
+deps =
+  pyopenssl
```
Instead of a dep list, could you add an `ssl` extra a few lines above?

jatalahd replied:
Done in latest push.

webknjaz commented on cherrypy/test/test_passwd.pem:
How difficult would it be to integrate `trustme` instead of committing test data into the repo?

jatalahd replied:
Done in latest push, but also reworked the whole test suite to use the more modern pytest.fixture approach. See what you think about it.

webknjaz commented on cherrypy/test/test_config_ssl_adapter.py:
```diff
@@ -0,0 +1,117 @@
+                    thisDir,
+                    'test_passwd.pem',
+                ),
+                'server.ssl_private_key_password': '123456',
```
Looking at this, I'm starting to question the interface — keeping secrets in long-living objects might not be the most secure decision (and decreasing security would probably defeat the purpose of putting passwords on keys).
I think it might be a good idea to allow this to also be a callable, not just a string. So that the end-users would be able to pass something that would retrieve the secret from a secure location. A function with a signature like `() -> str`. WDYT?

webknjaz replied:
A more concrete suggestion: https://github.com/cherrypy/cheroot/pull/798/files#r2566163502

jatalahd replied:
The cheroot commit is now merged, so I am not sure if anything needs to be done in cherrypy side?

webknjaz commented on cherrypy/_cpwsgi_server.py:
Looks like the same needs to happen here: https://github.com/cherrypy/cherrypy/blob/1f75bc9eed8e0e385f64f368bd69f58d96fb8c2b/cherrypy/_cpnative_server.py#L170-L187

jatalahd replied:
Done in latest push.

jatalahd commented:
@webknjaz ; In the latest push, reworked the whole test suite to use the pytest.fixture approach and using parametrization, added tests for both native and wsgi servers with pyopenssl and builtin adapters. There is quite a lot of code to review in the test suite, but please provide comments to make it better.

Also for the native server, the existing tests had excluded win32 platform, but I did not do that for the new tests. If this needs to be considered, please let me know.

webknjaz commented:
@jatalahd Thanks for working on this but I had a few health incidents over the past days that force me to stay away from GH for a while. Please keep iterating where you can, but JFYI it may take a while until I'll be able to get to another round of review.
